File: src/lib/invite.ts
/**
 * Group invite codes, and the links a QR code carries.
 *
 * The QR encodes a full URL rather than the bare code, so a phone's own
 * camera app can open it directly — someone without Snatzee installed
 * lands on the app instead of staring at a code they cannot use. The
 * in-app scanner accepts either shape.
 */

/** The link a QR code points at. Follows the domain it was generated on. */
export function inviteUrl(code: string, origin?: string) {
  const base = origin ?? (typeof window !== 'undefined' ? window.location.origin : '')
  return `${base}/app/groups?code=${encodeURIComponent(code)}`
}

/** Codes are 8 uppercase hex characters, but be generous about what is accepted. */
function normalise(value: string) {
  const code = value.trim().toUpperCase().replace(/[^A-Z0-9]/g, '')
  return /^[A-Z0-9]{4,16}$/.test(code) ? code : null
}

/**
 * Turns whatever was scanned or pasted into an invite code.
 *
 * Accepts a full invite URL, a bare code, or a code with stray spacing or
 * dashes in it. Returns null when it is neither.
 */
export function parseInviteCode(raw: string): string | null {
  const text = raw.trim()
  if (!text) return null

  if (/^https?:\/\//i.test(text)) {
    try {
      const fromQuery = new URL(text).searchParams.get('code')
      return fromQuery ? normalise(fromQuery) : null
    } catch {
      return null
    }
  }

  return normalise(text)
}

File: ios/Snatzee/Core/Invite.swift
import Foundation

/// Group invite codes and the links their QR codes carry — src/lib/invite.ts.
///
/// The QR encodes a full website URL rather than the bare code, so the
/// phone's own camera app can open it too (a universal link into this app,
/// or the website without it). The in-app scanner accepts either shape.
enum Invite {
    /// The link a QR code points at.
    static func url(code: String) -> URL? {
        var components = URLComponents()
        components.scheme = AppConfig.apiURL?.scheme ?? "https"
        components.host = AppConfig.apiURL?.host ?? "www.snatzee.nl"
        components.port = AppConfig.apiURL?.port
        components.path = "/app/groups"
        components.queryItems = [URLQueryItem(name: "code", value: code)]
        return components.url
    }

    /// Turns whatever was scanned or typed into an invite code, or nil.
    static func parse(_ raw: String) -> String? {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return nil }

        if text.range(of: "^https?://", options: [.regularExpression, .caseInsensitive]) != nil {
            guard let code = URLComponents(string: text)?.queryItems?.first(where: { $0.name == "code" })?.value else {
                return nil
            }
            return normalise(code)
        }
        return normalise(text)
    }

    private static func normalise(_ value: String) -> String? {
        let code = value.uppercased().filter { $0.isASCII && ($0.isLetter || $0.isNumber) }
        return (4...16).contains(code.count) ? code : nil
    }
}
